File: Java/LinkedList/01_Singly_Linear_LinkedList_Java/SinglyLL_009.java
/*
===============================================================================
Program Name    : Singly Linear Linked List - InsertFirst() and InsertLast()
Program Number  : 009
Description     : This program demonstrates the implementation of
                  InsertFirst() and InsertLast() operations in a
                  Singly Linear Linked List using Java.

                  The linked list is implemented using an
                  object-oriented approach where all data members
                  and member functions are encapsulated inside
                  the SinglyLL class.
                  
Date            : July 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve

===============================================================================
Concepts Used:
1. Classes and Objects
2. Self-Referential Class
3. Constructors
4. Object References
5. Encapsulation
6. Private Data Members
7. Public Member Functions
8. Dynamic Object Creation
9. InsertFirst()
10. InsertLast()

Key Learning Points:

- Linked list operations are implemented as member functions.
- first stores the reference of the first node.
- iCount stores the total number of nodes.
- InsertFirst() inserts a node at the beginning.
- InsertLast() inserts a node at the end.
- Constructors initialize newly created nodes automatically.
- Encapsulation improves code organization and security.

===============================================================================
*/
class node
{
    public int data;
    public node next;

    node(int no)
    {
        this.data = no;
        this.next = null;
    }
}

class SinglyLL
{
    private node first;
    private int iCount;

    public SinglyLL()
    {
        this.first = null;
        this.iCount = 0;
    }

    public void Display()
    {}

    public int Count()
    {
        return iCount;
    }

    public void InsertFirst(int iNo)
    {
        node newn = new node(iNo);
        newn.data = iNo;
        newn.next = null;

        if(first == null)
        {
            first = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }
    }

    public void InsertLast(int No)
    {
        node newn = new node(No);
        node temp = null;

        newn.data = No;
        newn.next = null;

        if(first == null)
        {
            first = newn;
        }
        else
        {
            temp = first;

            while(temp.next != null)
            {
                temp = temp.next;
            }

            temp.next = newn;
        }
    }

    public void InsertAtPos(int iNo,int iPos)
    {}

    public void DeleteFirst()
    {}

    public void DeleteLast()
    {}

    public void DeleteAtPos(int iPos)
    {}
}

class SinglyLL009
{
    public static void main(String A[])
    {
        SinglyLL sobj = new SinglyLL();

        sobj.InsertFirst(51);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);

        sobj.InsertLast(101);
        sobj.InsertLast(111);
        sobj.InsertLast(121);
    }
}
/*
===============================================================================
Expected Output
===============================================================================
No Output

The program creates a Singly Linear Linked List and performs
InsertFirst() and InsertLast() operations internally.

===============================================================================
Memory Diagram
===============================================================================

Initial List

first
  |
  v
null

After InsertFirst(51)

first
  |
  v
+------+------+
|  51  | null |
+------+------+

After InsertFirst(21)

first
  |
  v
+------+------+
|  21  |   o------+
+------+------+    |
                  v
             +------+------+
             |  51  | null |
             +------+------+

After InsertFirst(11)

first
  |
  v
+------+------+
|  11  |   o------+
+------+------+    |
                  v
             +------+------+
             |  21  |   o------+
             +------+------+    |
                               v
                          +------+------+
                          |  51  | null |
                          +------+------+

After InsertLast(101), InsertLast(111), InsertLast(121)

first
  |
  v
[11] → [21] → [51] → [101] → [111] → [121] → null

===============================================================================
Pointer / Logic Explanation
===============================================================================

InsertFirst()

Step 1:
Create a new node.

Step 2:
If the linked list is empty,
make the new node the first node.

Step 3:
Otherwise,

newn.next = first;
first = newn;

The new node becomes the first node.

------------------------------------------------------------

InsertLast()

Step 1:
Create a new node.

Step 2:
If the linked list is empty,
make the new node the first node.

Step 3:
Otherwise,

Traverse the linked list using a temporary
reference until the last node is reached.

temp.next = newn;

The new node becomes the last node.

===============================================================================
Operations Performed
===============================================================================
1. Created the node class.
2. Created the SinglyLL class.
3. Declared first as the head reference.
4. Declared iCount as the node counter.
5. Implemented Display() function prototype.
6. Implemented Count() function.
7. Implemented InsertFirst().
8. Implemented InsertLast().
9. Declared InsertAtPos().
10. Declared DeleteFirst().
11. Declared DeleteLast().
12. Declared DeleteAtPos().
13. Created a SinglyLL object.
14. Inserted nodes at the beginning.
15. Inserted nodes at the end.

===============================================================================
Time Complexity
===============================================================================
InsertFirst()      : O(1)

InsertLast()       : O(n)

Count()            : O(1)

Space Complexity   : O(n)

===============================================================================
Conclusion
===============================================================================
This program demonstrates the implementation of the
InsertFirst() and InsertLast() operations using an
object-oriented approach in Java. The linked list is
managed through the SinglyLL class, where both data
members and member functions are encapsulated together.

This program forms the foundation for implementing
Display(), Delete(), Search(), and other linked list
operations in the upcoming programs.

===============================================================================
*/

File: C++/LinkedList/01_Singly_Linear_LinkedList_CPP/SinglyLL_009.cpp
/*
===============================================================================
Program Name    : Display Operation in Singly Linked List
Program Number  : 009 
Description     : This program demonstrates the Display() operation of a
                  Singly Linked List. The Display() function traverses the
                  linked list from the first node and prints all the node
                  elements. Since no nodes are inserted, the linked list
                  remains empty and only NULL is displayed.
Date            : July 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Object-Oriented Programming (OOP)
2. Class and Object
3. Constructor
4. Self-Referential Structure
5. Linked List Traversal
6. Temporary Pointer
7. Encapsulation
8. Display Operation

Key Learning Points:
- Display() traverses the linked list from first to NULL.
- A temporary pointer is used for traversal.
- The original first pointer remains unchanged.
- An empty linked list is handled safely.
- Traversal continues until temp becomes NULL.

===============================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

class SinglyLL
{
    private:

        PNODE first;
        int iCount;

    public:

        SinglyLL();

        void Display();
        int Count();

        void InsertFirst(int iNo);
        void InsertLast(int iNo);
        void InsertAtPos(int iNo,int iPos);

        void DeleteFirst();
        void DeleteLast();
        void DeleteAtPos(int iPos);
};

/*
===============================================================================
Constructor Definition
===============================================================================
*/

SinglyLL::SinglyLL()
{
    this->first = NULL;
    this->iCount = 0;
}

/*
===============================================================================
Member Function Definitions
===============================================================================
*/

void SinglyLL::Display()
{
    PNODE temp = NULL;

    temp = this->first;

    while(temp != NULL)
    {
        cout<<"| "<<temp->data<<" | -> ";
        temp = temp->next;
    }

    cout<<"NULL"<<endl;
}

int SinglyLL::Count()
{
    return this->iCount;
}

void SinglyLL::InsertFirst(int iNo)
{
}

void SinglyLL::InsertLast(int iNo)
{
}

void SinglyLL::InsertAtPos(int iNo,int iPos)
{
}

void SinglyLL::DeleteFirst()
{
}

void SinglyLL::DeleteLast()
{
}

void SinglyLL::DeleteAtPos(int iPos)
{
}

int main()
{
    SinglyLL sobj;

    sobj.Display();

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

NULL

===============================================================================
Memory Representation:
===============================================================================

Linked List:

first
  │
  ▼
NULL

temp
 │
 ▼
NULL

===============================================================================
Explanation:
===============================================================================

The constructor initializes the linked list by setting first to NULL and
iCount to 0.

The Display() function creates a temporary pointer named temp and assigns it
the address stored in first.

Since no nodes have been inserted into the linked list, first contains NULL.
Therefore, temp also becomes NULL.

The while loop checks whether temp is not NULL. Since the condition is false,
the loop is skipped and only NULL is displayed.

Using a temporary pointer ensures that the original first pointer remains
unchanged after traversal.

===============================================================================
Program Flow:
===============================================================================

Create SinglyLL Object
          │
          ▼
Constructor Invoked
          │
          ▼
Initialize first = NULL
Initialize iCount = 0
          │
          ▼
Call Display()
          │
          ▼
temp = first
          │
          ▼
temp == NULL ?
          │
          ▼
Display NULL
          │
          ▼
Program Terminates

===============================================================================
Operations Performed:
===============================================================================

1. Defined a self-referential node structure.
2. Created the SinglyLL class.
3. Initialized first to NULL.
4. Initialized iCount to 0.
5. Created a temporary pointer for traversal.
6. Checked whether the linked list is empty.
7. Displayed NULL for an empty linked list.

===============================================================================
Time Complexity:
===============================================================================

Display()        : O(n)
Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 009 demonstrates the Display() operation of a Singly Linked List.
The function traverses the list using a temporary pointer and safely handles
the empty list condition by displaying NULL. This forms the basis for
displaying node elements after insertion operations in later programs.

===============================================================================
*/
